src/libstd/net/tcp.rs

```diff
@@ -0,0 +1,792 @@
+/// use std::net::TcpStream;
+///
+/// {
+///     let mut stream = TcpStream::connect("127.0.0.1:34254").unwrap();
+///
+///     // ignore the Result
+///     let _ = stream.write(&[1]);
+///     let _ = stream.read(&mut [0; 128]); // ignore here too
+/// } // the stream is closed here
+/// ```
+pub struct TcpStream(net_imp::TcpStream);
+
+/// A structure representing a socket server.
+///
+/// # Examples
+///
+/// ```no_run
+/// use std::net::{TcpListener, TcpStream};
+/// use std::thread::Thread;
+///
+/// let listener = TcpListener::bind("127.0.0.1:80").unwrap();
+///
+/// fn handle_client(stream: TcpStream) {
+///     // ...
+/// }
+///
+/// // accept connections and process them, spawning a new thread for each one
+/// for stream in listener.incoming() {
+///     match stream {
+///         Ok(stream) => {
+///             Thread::spawn(move|| {
+///                 // connection succeeded
+///                 handle_client(stream)
+///             });
+///         }
+///         Err(e) => { /* connection failed */ }
+///     }
+/// }
+///
+/// // close the socket server
+/// drop(listener);
+/// ```
+pub struct TcpListener(net_imp::TcpListener);
+
+/// An infinite iterator over the connections from a `TcpListener`.
+///
+/// This iterator will infinitely yield `Some` of the accepted connections. It
+/// is equivalent to calling `accept` in a loop.
+pub struct Incoming<'a> { listener: &'a TcpListener }
+
+impl TcpStream {
+    /// Open a TCP connection to a remote host.
+    ///
+    /// `addr` is an address of the remote host. Anything which implements
+    /// `ToSocketAddrs` trait can be supplied for the address; see this trait
+    /// documentation for concrete examples.
+    pub fn connect<A: ToSocketAddrs + ?Sized>(addr: &A) -> io::Result<TcpStream> {
+        super::each_addr(addr, net_imp::TcpStream::connect).map(TcpStream)
+    }
+
+    /// Returns the socket address of the remote peer of this TCP connection.
+    pub fn peer_addr(&self) -> io::Result<SocketAddr> {
+        self.0.peer_addr()
+    }
+
+    /// Returns the socket address of the local half of this TCP connection.
+    pub fn socket_addr(&self) -> io::Result<SocketAddr> {
+        self.0.socket_addr()
+    }
+
+    /// Shut down the read, write, or both halves of this connection.
+    ///
+    /// This function will cause all pending and future I/O on the specified
+    /// portions to return immediately with an appropriate value (see the
+    /// documentation of `Shutdown`).
+    pub fn shutdown(&self, how: Shutdown) -> io::Result<()> {
+        self.0.shutdown(how)
+    }
+
+    /// Create a new independently owned handle to the underlying socket.
+    ///
+    /// The returned `TcpStream` is a reference to the same stream that this
+    /// object references. Both handles will read and write the same stream of
+    /// data, and options set on one stream will be propagated to the other
+    /// stream.
+    pub fn duplicate(&self) -> io::Result<TcpStream> {
```

Comment:
I wonder if we should bikeshed this a bit more; seems a bit confusing to have alongside `clone`, and I suspect cases similar to this will crop up. Would it be too confusing to simply call this `clone` (and have the return type do the interesting work)?